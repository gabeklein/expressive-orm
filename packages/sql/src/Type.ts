import { Knex } from 'knex';
import { Connection } from './connection/Connection';
import { FIELD, Field } from './Field';
import { Query } from './Query';

export type InstanceOf<T> = T extends { prototype: infer U } ? U : never;

const REGISTER = new Map<Type.EntityType, Map<string, Field.Defined>>();
const CONNECTION = new Map<typeof Type, Connection>();

const describe = Object.getOwnPropertyDescriptor;
const define = Object.defineProperty;

declare namespace Type {
  type EntityType<T extends Type = Type> =
    & (abstract new () => T)
    & typeof Type;

  type List<T extends Type> = T[];

  type DataRecusive<T> =
    T extends string ? string :
    T extends number ? number :
    T extends boolean ? boolean :
    T extends List<infer U> ? Pure<U>[] :
    T extends Type ? Pure<T> : T;

  type Pure<T extends Type> = {
    [K in Field<T>]-?: DataRecusive<T[K]>;
  }

  type Field<T extends Type> = Exclude<keyof T, "table">;

  namespace Insert {
    type Property<T> = T extends Type ? T | number : T;
  }

  type Insert<T extends Type> = {
    [K in Type.Field<T>]?: Insert.Property<T[K]>;
  }

  type QueryFunction<T extends Type, R> =
    (query: Query.FromType<T>, where: Query.Where) => R;
}

abstract class Type {
  // why is this non-nullable?
  this!: Type.EntityType;

  /**
   * Primary key of this entity.
   * May be any name in the actual database, however requred to be `id` as property of this type.
   * 
   * Setting to zero will disable this field, and ORM will not expect it to be present in the database.
   */
  id: number | string = Field({
    datatype: "int",
    nullable: false,
    increment: true,
    primary: true
  });

  static table = "";
  static schema = "";
  static deps = new Set<Type.EntityType>();

  static focus?: { [key: string]: any };

  static get connection(){
    return CONNECTION.get(this) || Connection.default;
  }

  static set connection(connection: Connection){
    CONNECTION.set(this, connection);
  }

  static toString(){
    return this.name;
  }

  static get fields(){
    return REGISTER.get(this) || init(this);
  }

  static digest<T extends Type>(data: Type.Insert<T>): Record<string, unknown>;
  static digest<T extends Type>(data: Type.Insert<T>[]): Record<string, unknown>[];
  static digest<T extends Type>(data: Type.Insert<T> | Type.Insert<T>[]){
    return Array.isArray(data) ? data.map(sanitize, this) : sanitize.call(this, data);  
  }

  static insert<T extends Type>(this: Type.EntityType<T>, data: Type.Insert<T> | Iterable<Type.Insert<T>>): Promise<void>;
  static insert<T extends Type>(this: Type.EntityType<T>, number: number, mapper: (index: number) => Type.Insert<T>): Promise<void>;
  static insert<T extends Type>(this: Type.EntityType<T>, data: Type.Insert<T> | Iterable<Type.Insert<T>> | number, map?: (i: number) => Type.Insert<T>){
    if(typeof data == "number"){
      if(!map)
        throw new Error("Cannot insert a number without a map function!");
  
      data = Array.from({ length: data }, (_, i) => map(i));
    }
    else if(!isIterable(data))
      data = [data];

    if(!this.connection)
      throw new Error("No connection found for type");

    const rows = Array.from(data).map(sanitize, this);
    const table = this.connection.knex(this.table);
    
    return table.insert(rows) as Knex.QueryBuilder;
  }

  static get<T extends Type, R>(
    this: Type.EntityType<T>,
    query: Type.QueryFunction<T, R>){

    return Query(where => {
      return query(where(this), where);
    });
  }
}

function init(type: Type.EntityType){
  const fields = new Map();

  if(!type.table)
    type.table = type.name.replace(/^[A-Z]/, m => m.toLowerCase());

  REGISTER.set(type, fields);
  
  const reference = new (type as any)();
  
  for(const key in reference){
    const { value } = describe(reference, key)!;
    const instruction = FIELD.get(value);    

    if(!instruction)
      throw new Error(`Entities do not support normal values, only fields. Did you forget to import \`${capitalize(typeof value)}\`?`);

    FIELD.delete(value);
    instruction(type, key);
    define(reference, key, {
      get: () => type.focus![key],
      set: is => type.focus![key] = is
    })
  }

  return fields;
}

function sanitize<T extends Type>(
  this: Type.EntityType<T>,
  data: Type.Insert<T>,
  index?: number,
  array?: Type.Insert<T>[]){

  const values: Record<string, any> = {};

  this.fields.forEach((field, key) => {
    const given = data[key as Type.Field<T>] as unknown;
    const which = array && array.length > 1 ? ` at [${index}]` : ""

    if(given == null){
      if(field.nullable || field.default || field.primary)
        return

      throw new Error(
        `Provided value for ${this.name}.${key} is ${given}${which} but column is not nullable.`
      );
    }

    try {
      const value = field.set ? field.set(given) : undefined;
      values[field.column] = value === undefined ? given : value;
    }
    catch(err: unknown){
      let message = `Provided value for ${this.name}.${key}${which} but not acceptable for type ${field.datatype}.`;

      if(typeof err == "string")
        message += `\n${err}`;

      throw err instanceof Error ? err : new Error(message);
    }
  });

  return values;
}

function isIterable(obj: unknown): obj is Iterable<unknown> {
  type MaybeIterable = {
    [Symbol.iterator]?: () => IterableIterator<unknown>;
  };

  return obj != null && typeof (obj as MaybeIterable)[Symbol.iterator] === 'function';
}

export function isTypeConstructor(obj: unknown): obj is typeof Type {
  return typeof obj === 'function' && obj.prototype instanceof Type;
}

export function capitalize(string: string){
  return string[0].toUpperCase() + string.slice(1);
}

export { Type }