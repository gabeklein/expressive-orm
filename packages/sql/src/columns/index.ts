export { Bool } from './Bool';
export { Num } from './Num';
export { One } from './One';
export { Str } from './Str';
export { Time } from './Time';