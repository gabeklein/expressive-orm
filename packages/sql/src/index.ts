export { Bool } from './field/Bool';
export { Connection } from './connection/Connection';
export { default as Util } from './utility';
export { Enum } from './field/Enum';
export { Field } from './Field';
export { Join } from './field/Join';
export { Json } from './field/Json';
export { Many } from './field/Many';
export { Num } from './field/Num';
export { One } from './field/One';
export { Query } from './query/Query';
export { Ref } from './field/Ref';
export { Scanner } from './parse/Scanner';
export { Schema } from './connection/Schema';
export { Str } from './field/Str';
export { Table } from './field/Table';
export { Time } from './field/Time';
export { Type } from './Type';