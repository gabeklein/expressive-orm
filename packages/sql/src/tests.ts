import { Connection } from ".";

const reset = new Set<Function>();
let cleanup: Function | undefined;

afterEach(() => {
  return Promise.all(Array.from(reset).map(cb => {
    reset.delete(cb);
    return cb();
  }));
});

afterAll(() => cleanup && cleanup());

/**
 * An in-memory database specific to a
 * test and attaches entities provided to it.
 **/
export class TestConnection extends Connection {
  constructor(
    public using: Connection.Types,
    private after?: () => void){

    super(using, {
      client: "sqlite3",
      useNullAsDefault: true,
      connection: {
        filename: ":memory:"
      }
    });

    if(expect.getState().currentTestName)
      reset.add(() => this.close());
    else {
      beforeAll(async () => this);
      cleanup = () => this.close();
    }
  }

  then(onfulfilled?: (value: void) => any) {
    this.sync().then(() => {
      if(this.after) this.after();
      if(onfulfilled) onfulfilled();
    });
  }
}