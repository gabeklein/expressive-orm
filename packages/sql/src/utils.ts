export function capitalize(string: string){
  return string[0].toUpperCase() + string.slice(1);
}

export function isIterable(obj: unknown): obj is Iterable<unknown> {
  type MaybeIterable = {
    [Symbol.iterator]?: () => IterableIterator<unknown>;
  };

  return obj != null && typeof (obj as MaybeIterable)[Symbol.iterator] === 'function';
}

export function underscore(str: string){
  return str.replace(/([a-z])([A-Z])/g, "$1_$2").toLowerCase();
}