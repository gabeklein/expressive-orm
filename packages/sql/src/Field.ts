import { RelevantTable } from './Query';
import { Type } from './Type';

export const FIELD = new Map<symbol, Field.Init>();

declare namespace Field {
  type Defined = {
    readonly [K in keyof Field]-?: Field[K];
  }

  type Init = (parent: Type.EntityType, key: string) => void;

  const defaults: Partial<Field>;

  function is(value: unknown): value is Defined;
}

interface Field {
  column?: string;
  datatype?: string;
  unique?: boolean;
  nullable?: boolean;
  primary?: boolean;
  increment?: boolean;
  default?: string | null;

  index?: number;

  /** Converts acceptable values to their respective database values. */
  set?(value: unknown): any;

  /** Converts database values to type of value in javascript. */
  get?(value: unknown): any;
}

function Field(options: Partial<Field> | Field.Init){
  const placeholder = Symbol(`field`);

  if(typeof options == "object"){
    const inputs = options;

    options = (parent, key) => {
      const field = Object.create(Field.prototype);
  
      Object.assign(field, Field.defaults, inputs, {
        column: inputs.column || key,
        toString(this: Field){
          const table = RelevantTable.get(this);
          const prefix = table ? `${table.alias || table.name}.` : "";
  
          return prefix + this.column;
        }
      });
  
      Object.freeze(field);
      parent.fields.set(key, field);
  
      return field;
    }
  }

  FIELD.set(placeholder, options);
  
  return placeholder as any;
}

Object.defineProperty(Field, "is", {
  writable: false,
  value(value: unknown){
    return value instanceof Field;
  },
})

Object.defineProperty(Field, "defaults", {
  writable: false,
  value: <Field.Defined>{
    column: "",
    index: 0,
    datatype: "varchar(255)",
    nullable: false,
    unique: false,
    primary: false,
    increment: false,
    default: null,
    set: x => x,
    get: x => x,
  }
})

export { Field }