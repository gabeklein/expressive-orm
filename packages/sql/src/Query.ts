import knex, { Knex } from 'knex';

import { Field } from './Field';
import { isTypeConstructor, Type } from './Type';

export const RelevantTable = new WeakMap<{}, Query.Table>();
declare const ENTITY: unique symbol;

declare namespace Query { 
  namespace Join {
    type Mode = "left" | "inner";

    type Where = (field: Field) => {
      is(equalTo: Field): void;
      isNot(equalTo: Field): void;
      isMore(than: Field): void;
      isLess(than: Field): void;
    }

    type Function = (on: Where) => void;

    type Object<T extends Type> = {
      [K in Type.Field<T>]?: T[K];
    }
  }

  type Assertion = symbol;

  interface Compare<T> {
    is(equalTo: T): Assertion;
    isNot(equalTo: T): Assertion;
    isMore(than: T): Assertion;
    isLess(than: T): Assertion;
  }

  interface Assert<T> extends Compare<T> {
    isAsc(): void;
    isDesc(): void;
  }

  interface Verbs <T extends Type> {
    delete(): void;
    update(values: Query.Update<T>): void;
  }

  interface Table {
    type: Type.EntityType;
    name: string | Knex.AliasDict;
    alias?: string;
  }

  type FromType<T extends Type = Type> = {
    [K in Type.Field<T>]: Exclude<T[K], null>;
  } & {
    [ENTITY]?: T
  }

  type Compat<T extends Type = Type> = {
    [K in Type.Field<T>]?: T[K];
  }

  // TODO: make this default/nullable aware.
  type Update<T extends Type> = {
    [K in Type.Field<T>]?: Exclude<T[K], undefined>;
  }

  interface Where {
    <T extends Type>(entity: Type.EntityType<T>, on?: Compat<T> | Join.Function, join?: "inner"): FromType<T>;
    <T extends Type>(entity: Type.EntityType<T>, on: Compat<T> | Join.Function, join: Join.Mode): Partial<FromType<T>>;

    <T extends Type>(field: FromType<T>): Verbs<T>;
    
    <T>(field: T): Assert<T>;
  }

  type Function<R> = (where: Where) => R;
}

interface Query<T = unknown> extends PromiseLike<T> {
  /**
   * Counts the number of rows that would be selected or modified.
   * Will ignore any side effects, such as update and delete.
   **/
  count(): Promise<number>;

  /**
   * Returns the SQL string generated by this query.
   */
  toString(): string;
}

interface SelectQuery<T = unknown> extends Query<T[]> {
  /**
   * Returns the first row that matches creteria.
   * 
   * @param orFail If true, will throw an error if no results are returned.
   */
  one(orFail?: boolean): Promise<T>;
}

/** Executable query on whole types. */
interface TypeQuery<T extends Type = any> extends SelectQuery<T> {
  update(values: Type.Insert<T>): Promise<number>;
  delete(): Promise<number>;
}

function Query<T extends Type>(from: Query.Function<T>): TypeQuery<T>;

function Query<T extends {}>(from: Query.Function<T>): SelectQuery<T>;

/**
 * Creates a new query.
 * 
 * If no selection is returned by the constructor, will return
 * the number of rows that would be selected or modified.
 */
function Query(from: Query.Function<void>): Query<number>;

function Query<T = void>(constructor: Query.Function<T>): Query | SelectQuery | TypeQuery {
  const tables = [] as Query.Table[];
  const pending = new Set<() => void>();

  let parse = (raw: unknown[]) => raw;
  let builder!: Knex.QueryBuilder;
  let main: Type.EntityType;

  const selects = constructor(where);

  pending.forEach(fn => fn());
  pending.clear();

  if(selects){
    builder.clearSelect();

    if(Field.is(selects)){
      builder.select({
        [selects.column]: String(selects)
      });
  
      parse = raw => raw.map(row => {
        const value = (row as any)[selects.column];
        return selects.get ? selects.get(value) : value;
      });
    }
    else if(typeof selects == "object"){
      const output = new Map<string | Field, string | number>();

      Object.getOwnPropertyNames(selects).forEach(key => {
        const value = (selects as any)[key];
    
        if(Field.is(value)){
          output.set(value, key);
          builder.select({ [key]: String(value) });
        }
      })

      parse = raw => raw.map(row => {
        const values = Object.create(selects as {});
    
        output.forEach((column, value) => {
          if(Field.is(value) && value.get)
            value = value.get((row as any)[column]);

          Object.defineProperty(values, column, { value });
        })
        
        return values as T;
      })
    }
  }

  function where(
    type: unknown,
    on?: Query.Compat | Query.Join.Function,
    join?: Query.Join.Mode
  ): any {
    if(isTypeConstructor(type)){
      if(typeof on == "string")
        throw new Error("Bad parameters.");

      return table(type, on, join);
    }

    if(Field.is(type)){
      const ref = String(type);
      const compare = (op: string) => (right: any) => {
        if(typeof right !== "number")
          right = String(right);

        builder.where(ref, op, right);
      }
  
      return {
        is: compare("="),
        isNot: compare("<>"),
        isMore: compare(">"),
        isLess: compare("<"),
        isAsc: () => builder.orderBy(ref, "asc"),
        isDesc: () => builder.orderBy(ref, "desc"),
      }
    }

    if(isFromType(type))
      return verbs(type);

    throw new Error("Invalid query.");
  }

  function verbs<T extends Type>(from: Query.FromType<T>): Query.Verbs<T> {
    const table = RelevantTable.get(from);

    if(!table)
      throw new Error(`Argument ${from} is not a query entity.`);

    return {
      delete: () => {
        builder.table(table.name).delete();
      },
      update: (data: Query.Update<any>) => {
        data = table.type.digest(data);
        builder.table(table.name).update(data);
      }
    }
  }

  function table(
    type: Type.EntityType,
    on?: Query.Compat | Query.Join.Function,
    mode?: Query.Join.Mode
  ): Query.FromType {
    let { fields, schema } = type;
    let name: string | Knex.AliasDict = type.table
    let alias: string | undefined;

    if(schema){
      alias = `$${tables.length}`;
      name = { [alias]: schema + '.' + name };
    }

    const proxy = {} as any;
    const metadata: Query.Table = { name, alias, type };

    tables.push(metadata);
    RelevantTable.set(proxy, metadata);

    fields.forEach((field, key) => {
      field = Object.create(field);
      RelevantTable.set(field, metadata);
      Object.defineProperty(proxy, key, { value: field })
    })

    if(main === undefined){
      main = type;

      const engine = type.connection?.knex || knex({
        client: "sqlite3",
        useNullAsDefault: true,
        pool: { max: 0 }
      });

      builder = engine(name).select({ count: "COUNT(*)" });
    }
    else {
      if(type.connection !== main.connection)
        throw new Error(`Joined entity ${type} does not share a connection with ${main}`);

      let callback: Knex.JoinCallback;

      switch(typeof on){
        case "function":
          callback = (table) => {
            pending.add(() => {
              on(field => {
                if (!Field.is(field))
                  throw new Error("Join assertions can only apply to fields.");

                const on = (op: string) => (right: Field) => {
                  table.on(String(field), op, String(right));
                };

                return {
                  is: on("="),
                  isNot: on("<>"),
                  isMore: on(">"),
                  isLess: on("<"),
                }
              });
            })
          }
        break;

        case "object":
          callback = (table) => {
            for (const key in on) {
              const field = type.fields.get(key);
    
              if (!field)
                throw new Error(`${key} is not a valid field in ${type}.`);
    
              const left = `${type.table}.${field.column}`;
              const right = String((on as any)[key]);
    
              table.on(left, "=", right);
            }
          }
        break;

        default:
          throw new Error(`Invalid join on: ${on}`);
      }

      switch(mode){
        case "left":
          builder.leftJoin(name, callback);
          break;
  
        case "inner":
        case undefined:
          builder.join(name, callback);
          break;
  
        case "right" as unknown:
        case "full" as unknown:
          throw new Error(`Cannot ${mode} join because that would affect ${main} which is already defined.`);
  
        default:
          throw new Error(`Invalid join type ${mode}.`);
      }
    }

    return proxy;
  }

  const query: Query = {
    then(resolve: (res: any) => any, reject: (err: any) => any){
      return builder.then(parse).then<T[]>(resolve).catch(reject);
    },
    count(){
      return builder.clone().clearSelect().count();
    },
    toString(){
      return builder.toString().replace(/```/g, "`");
    }
  }

  if(selects)
    return {
      ...query,
      async one(orFail?: boolean){
        const res = await builder.clone().limit(1).then(parse);
    
        if(res.length == 0 && orFail)
          throw new Error("Query returned no results.");
    
        return res[0] as T;
      }
    }

  return query;
}

Query.one = function one<T extends {}>(
  where: Query.Function<T>, orFail?: boolean){

  return Query(where).one(orFail);
}

function isFromType(type: any): type is Query.FromType {
  return RelevantTable.has(type);
}

export { Query }