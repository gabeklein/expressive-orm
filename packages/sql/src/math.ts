import { Field } from "./Field";
import { Query } from "./Query";

type Value = Query.Value;
type ANumeric = Query.ANumeric;
type Numeric = Query.Numeric;

export interface Bitwise {
  not(value: Numeric): Numeric;
  and(left: Numeric, right: Numeric): Numeric;
  or(left: Numeric, right: Numeric): Numeric;
  xor(left: Numeric, right: Numeric): Numeric;
  left(value: Numeric, shift: Numeric): Numeric;
  right(value: Numeric, shift: Numeric): Numeric;
}

export interface MathOps {
  add(left: Numeric, right: Numeric): Numeric;
  add(left: ANumeric, right: ANumeric): ANumeric;
  sub(left: Numeric, right: Numeric): Numeric;
  mul(left: Numeric, right: Numeric): Numeric;
  div(left: Numeric, right: Numeric): Numeric;
  mod(left: Numeric, right: Numeric): Numeric;
  neg(value: Numeric): Numeric;
  pos(value: Numeric): Numeric;
  bit: Bitwise
}

export function math(): MathOps {
  function op(op: string, rank: number, unary: true): (value: Value) => Value;
  function op(op: string, rank: number, unary?: false): (left: Value, right: Value) => Value;
  function op(op: string, rank: number, arg2?: boolean){
    return (l: any, r?: any) => {
      const input = arg2 === true ? [op, l] : [l, op, r]
      const computed = new Computed(...input);

      computed.rank = rank || 0;

      return computed;
    }
  }

  return {
    add: op('+', 4),
    sub: op('-', 4),
    mul: op('*', 5),
    div: op('/', 5),
    mod: op('%,', 5),
    neg: op('-', 7, true),
    pos: op('+', 7, true),
    bit: {
      not: op('~', 6, true),
      left: op('<<', 3),
      right: op('>>', 3),
      and: op('&', 2),
      or: op('|', 0),
      xor: op('^', 1),
    }
  }
}

export class Computed<T> extends Array<T | Field<T> | Computed<T>> {
  rank = 0;

  get(input: unknown){
    return input as string;
  }
  
  toString(): string {
    return this.map(value => {
      if(typeof value == "number")
        return value;

      if(value instanceof Computed)
        return value.rank > this.rank ? value : `(${value})`;

      return String(value);
    }).join(" ");
  }
}