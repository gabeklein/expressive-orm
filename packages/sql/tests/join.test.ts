import { Entity, Num, Query, Str, Table } from '../src';

class Foo extends Entity {
  name = Str();
  color = Str();
}

class Bar extends Entity {
  name = Str();
  color = Str();
  rating = Num();
}

class Baz extends Entity {
  color = Str();
  rating = Num();
}

it("will join using object", async () => {
  const query = new Query(where => {
    const foo = where(Foo);
    const bar = where(Bar, { color: foo.color });
    const baz = where(Baz, { rating: bar.rating });

    where(foo.name).not("Danny");
    where(bar.rating).greater(50);
    where(baz.color).is("blue");
  });

  expect(query).toMatchSnapshot();
})

it("will join using function", async () => {
  const query = new Query(where => {
    const foo = where(Foo);
    const bar = where(Bar, on => {
      on(bar.name).not(foo.name);
      on(bar.color).is(foo.color);
    });

    where(foo.name).not("Danny");
    where(bar.rating).greater(50);
  });

  expect(query).toMatchSnapshot();
})

it("will alias tables which have a schema", () => {
  class Foo extends Entity {
    this = Table({ schema: "foobar" });

    name = Str();
    color = Str();
  }

  const query = new Query(where => {
    const foo = where(Foo);

    where(foo.color).is("red");
  })

  expect(query).toMatchSnapshot();
})