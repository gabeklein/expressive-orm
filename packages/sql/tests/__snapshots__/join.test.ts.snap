// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`will alias tables which have a schema 1`] = `
	select
	  \`COUNT(*)\` as \`count\`
	from
	  \`foobar\`.\`Foo\` as \`$0\`
	where
	  \`$0\`.\`color\` = 'red'
`;

exports[`will join using function 1`] = `
	select
	  \`COUNT(*)\` as \`count\`
	from
	  \`Foo\`
	  inner join \`Bar\` on \`Bar\`.\`name\` <> \`Foo\`.\`name\`
	  and \`Bar\`.\`color\` = \`Foo\`.\`color\`
	where
	  \`Foo\`.\`name\` <> 'Danny'
	  and \`Bar\`.\`rating\` > 50
`;

exports[`will join using object 1`] = `
	select
	  \`COUNT(*)\` as \`count\`
	from
	  \`Foo\`
	  inner join \`Bar\` on \`Bar\`.\`color\` = \`Foo\`.\`color\`
	  inner join \`Baz\` on \`Baz\`.\`rating\` = \`Bar\`.\`rating\`
	where
	  \`Foo\`.\`name\` <> 'Danny'
	  and \`Bar\`.\`rating\` > 50
	  and \`Baz\`.\`color\` = 'blue'
`;
